### ISSUE TITLE
`buck run` does not forward stdin

### ISSUE THREAD

**Issue by artemyarulin** [facebook/buck#824]
Hi,

I'm trying to create repl-like functionality for my targets using `genrule` with `executable=True` and noticed that standard input is not available, here some dummy example:

``` python
genrule(name = 'repl',
        srcs = [],
        bash = "echo 'echo Enter name: && read name && echo $name' > $OUT && chmod +x $OUT",
        executable = True,
        out = 'repl')
```

When I run it like `buck run //:repl` I have output with `Enter name:` and then I can type and press enter, but nothing happens, I guess input doesn't get redirected to runnable so it stuck waiting forever.

Is this by design or is it an issue?

Thanks

**Comment by mikekap** [comment]
I think this is a bad behavior with `buck run` in general - stdin isn't sent to the process. Running the output directly wouldn't have this behavior.

Also note that for that simple case, you could use `sh_binary`.

**Comment by marcinkosiba** [comment]
yea, the `buck run` command is broken: it runs the command as a background process and forwards stout/stderr but not stdin. It really should work by calling exec in the 'client' process instead.

That being said your use of `genrule` is incorrect. genrules are for running arbitrary commands to create artifacts that are consumed at later stages of the build. The inputs of those commands must be completely exposed to the build system, otherwise you get incorrect caching behavior.
In your example the result of `read` is not captured at all.

I'm going to re-purpose this issue to address the fact that 'buck run ignores stdin`. For the time being I suggest you build a standalone repl thing and invoke it from outside of the buck process:

`buck targets --show-output <name of target> | cut -f2 -d" "`

**Comment by artemyarulin** [comment]
Thank you all for your comments and suggestions. 

I've found a workaround for my case - Clojure repl has a network mode where it just starts listening on port for commands, so I'm fine with it for now.

If you'll need any testing in a feature - feel free to ping me. 

Thanks again for such an amazing peace of software, really happy with Buck

**Comment by bertmaher** [comment]
I've implemented this feature as of this commit: a271a23a24a63e88decba55e176d623717628719.  In addition to properly inheriting stdin, that diff will run commands entirely in the client, so the buckd lock isn't held for the duration of the run.
